keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+* `kube_pod_resource_requests_lifecycle` contains series for the consumed requested resources for a pod at a given time when the pod is in the provided lifecycle phase. All pods are in one of the lifecycle phases, but series will only be returned for the current phase.
+  * The value is the quantity of the limit in `spec.containers.resources.requests`
+  * It has the following labels:
+    * `lifecycle` - this is set to the current lifecycle phase as described in the resource model - `Pending`, `Initializing`, `Running`, `Completed`
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_limits_lifecycle` contains series for the consumed limit resources for a pod at a given time when the pod is in the provided lifecycle phase.
+All pods are in one of the lifecycle phases, but series will only be returned for the current phase.
+  * The value is the quantity of the limit in `spec.containers.resources.limits`
+  * It has the following labels:
+    * `lifecycle` - this is set to the current lifecycle phase as described in the resource model - `Pending`, `Initializing`, `Running`, `Completed`
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+
+This will add `O(pods * resource types)` metrics series, and in general most clusters have 4-8 resource types. The series per pod is `resource_count * 2 * 2 (requests and limits for normal and _lifecycle)`.  This explicitly does not expose containers as a dimension, as the resource model applies rules to containers that cannot be trivially expressed via a formula and does not help answer the core questions.
+
+These metrics should roughly correspond to:
+
+* The resources the scheduler will consider the pod to be consuming at the current time
+* The resources the kubelet will use to decide whether to admit the pod (nodes will reject requested resources that exceed their actual allocatable capacity)
+* The effective reserved capacity as resource quota would observe for the namespace
+* The sum of all resources consumed on the cluster
+* The size of cgroup constraints applied to pods on a node
+
+The lifecycle metrics allow key insights that are impossible today:
+
+* Observing when init containers or containers have substantially different resource reservation, causing workloads to under provision nodes
```

Comment:
The outcome isn't really under-provisioning, as the "pod requested" resources are still granted to containers.  I suppose for guaranteed pods this would cause under-provisioning, as containers' limits prevent them from using all the resources the pod requests.

It is more that we want to help users identify when they are under-utilizing resources granted to their pods.  IMO, it seems easy enough to do this by comparing `kube_pod_resource_requests` to measured usage, or to `sum(kube_pod_container_resource_requests)`.

Reply:
The difference between the phases is pretty significant, and you might have high request on initialization and low request after, or vice versa, and knowing your usage during those phases (for anything with non-trivial initialization) is pretty key to understanding where to change.  Init containers are used pretty heavily in the wild in job systems and I have seen plenty of places where visibility into the lifecycle was the key insight that allowed someone to adjust the job to match the request based on dimensions the pod model doesn't expose well (time and usage)

Reply:
I'm trying to figure out how you see the feedback loop to users working.

I think we are in agreement that this is primarily about giving users the tools to eliminate waste.  We are also in agreement that the current set of tools (just container-level metrics) are insufficient for most users to even be able to measure waste correctly.  Given `kube_pod_resource_requests`, that seems like it should be the right metric to compare against aggregated pod usage to measure utilization across time, regardless of which containers usage or requests are coming from.

If I want to answer "why are my requests so high", it seems difficult to do with metrics.  Maybe we have different intuitions, but I wouldn't intuit that a higher initializing value and a lower running value meant that I was wasting resources.  I would probably try comparing my usage to the metric, which would lead me to incorrect conclusions.  I would still have to go look up the docs to actually figure out how to solve my problem.  If I were designing a metric to help me find "wasteful" pods, one thing I would want to do would be to create an alert for it, or monitor the number of such pods across my cluster(s).  But to do that, I would want metric streams that overlap in time, so they can be compared.

Reply:
Ah, I was thinking more about having baked dashboard examples that cover this (ultimately this whole design leads to "I believe it should be possible to write an accurate capacity dashboard for Kubernetes, but it is not today").  So yes, there are multiple types of alerts and visualizations that would show this by default.

For instance, if you have a pod dashboard, I would replace pod use over time with 3 lines - one which is solid which is "the current usage", one which is light dashed "the requests", and one which is even lighter dashed which is "the phase".  If you look at any pod with meaningful init containers, you would see that distinction.  I would probably have a per cluster commit %, an overcommit %, a waste usage %, a phase specific waste usage %, and then highlight the most contended resources.

I would also have a couple of alerts potentially by default:

* If requested capacity exceeds total capacity on any resource dimension, alert
* If a workload significantly under requests, alert (possibly into a different context or make this optional, obviously this is highly specific, just thinking about what 99% of kube admins get wrong)
* if a workload significantly under requests BECAUSE of initialization, alert (same as above, maybe a different context)

Reply:
Ironically, at this very moment I am debugging a scheduler contention bug in an e2e test that runs serial where without this metric I don't know whether the initializers or containers are higher for a set of pods that ran over the whole test (where the full set of pods is basically unknowable since it's coming out of 200 e2e runs).  So concretely being able to debug where a scheduling constraint was hit, and then be able to overlay the distinction between resources requested during phase would eliminate the likely failure without having to know the full context of every pod.  I think that maps pretty clearly to an admin role in a similar mode.

Reply:
I don't _think_ we should encourage comparing the request lifecycle metric to usage, as that isn't necessarily what the pod is granted.

I can see wanting to compare the request lifecycle metric with phase=Running to `kube_pod_resource_requests`.  If they aren't the same, you know init containers are the "request-setting" piece.

The phase-specific waste usage % is an interesting thought, but I don't think it applies well to all the phases, as you only have actual usage for Initializing and Running.  Or would we measure Pending against Quota to measure waste?  For Initializing and Running, I think you would want to compare usage to `kube_pod_resource_requests`, rather than to the `_lifecycle` metric, since I care about my usage relative to what i'm allowed to use + what counts against my quota.  I can see wanting to break down the result by phase, but that would need a `kube_pod_resource_requests` with the lifecycle dimension instead of the proposed lifecycle metric...

Do you have KSM metrics in the e2e test?

> So concretely being able to debug where a scheduling constraint was hit, and then be able to overlay the distinction between resources requested during phase would eliminate the likely failure without having to know the full context of every pod

I'm not sure we can represent the host of reasons why scheduling constraints are hit with metrics easily.  But I agree we can at least know when we have more total pod requests than allocatable in a cluster.

Is this a "what is using up all my capacity" issue, a "what is using up all the resources in my cluster" issue, or a "why is my cluster utilization so low" issue (or something else)?  If it is the first, you would probably want to compare `kube_pod_resource_requests` vs total allocatable.  If the second, just look at pod usage.  If the third, I think you would just want to compare `kube_pod_resources_requests` to usage to identify the pods with the lowest utilization.

Reply:
After reflecting on this metric and actually going back to the implementation I realized what the minimal piece that is necessary to convey in the lifecycle metric - the distinction between pending scheduled and running resources.  While I really would prefer to be able to visualize mismatches between init container and container max, I think it's ok to pass on that for now and address it later.  But a user does need to be able to sum currently running resources as distinct from scheduled + running resources.

To that end, I'd propose we start with the following metrics groups:

```
kube_pod_resources* - resources of non-terminal pods
kube_running_pod_resources* - resources of non-terminal, non-pending pods
```

That allows someone to calculate:

* total running resources (sum running)
* total pending resources (sum pod and absent running)
* total desired workload (sum pod)

which was a key goal and is pretty common.

Reply:
That is reasonable to me.